src/Math/src/CARE.cpp

```diff
@@ -0,0 +1,191 @@
+                       Eigen::Ref<const Eigen::MatrixXd> R)
+{
+    constexpr std::string_view errorPrefix = "[CARE::setMatrices] ";
+
+    const std::size_t n = B.rows();
+    const std::size_t m = B.cols();
+
+    // check the size of the matrices
+    if (A.rows() != n || A.cols() != n)
+    {
+        std::cerr << errorPrefix << "The matrix A must be a square matrix. Expected size: " << n
+                  << " x " << n << ". Passed size: " << A.rows() << " x " << A.cols() << "."
+                  << std::endl;
+        return false;
+    }
+
+    if (Q.rows() != n || Q.cols() != n)
+    {
+        std::cerr << errorPrefix << "The matrix Q must be a square matrix. Expected size: " << n
+                  << " x " << n << ". Passed size: " << Q.rows() << " x " << Q.cols() << "."
+                  << std::endl;
+        return false;
+    }
+
+    if (R.rows() != m || R.cols() != m)
+    {
+        std::cerr << errorPrefix << "The matrix R must be a square matrix. Expected size: " << m
+                  << " x " << m << ". Passed size: " << R.rows() << " x " << R.cols() << "."
+                  << std::endl;
+        return false;
+    }
+
+    // this check may be time consuming let's consider it only when the library is compiled in debug
+    constexpr double symmetricTolerance = 1e-8;
+    assert(Q.isApprox(Q.transpose(), symmetricTolerance) && "The matrix Q must be symmetric");
+
+    m_pimpl->A = A;
+    m_pimpl->B = B;
+    m_pimpl->Q = Q;
+    m_pimpl->R = R;
+
+    m_pimpl->resizeInteralMatrices();
+
+    return true;
+}
+
+bool CARE::solve()
+{
+    // compute cholesky decomposition
+    Eigen::LLT<Eigen::MatrixXd> R_cholesky(m_pimpl->R);
+    if (R_cholesky.info() != Eigen::Success)
+    {
+        std::cerr << "[CARE::solve] The matrix R must be positive definite." << std::endl;
+        return false;
+    }
+
+    // Z represents the Hamiltonian matrix.
+    //       _                _
+    //      | A   B inv(R) B'  |
+    // Z  = |                  |
+    //      |_Q       -  A'   _|
+
+    m_pimpl->Z << m_pimpl->A, m_pimpl->B * R_cholesky.solve(m_pimpl->B.transpose()),
+                  m_pimpl->Q, -m_pimpl->A.transpose();
+
+    double relativeNorm = m_pimpl->tolerance;
+    std::size_t iteration = 0;
+    const double p = static_cast<double>(m_pimpl->Z.rows());
+
+    // run the algorithm
+    for (; iteration < m_pimpl->maxIterations && relativeNorm >= m_pimpl->tolerance; iteration++)
+    {
+        m_pimpl->ZOld = m_pimpl->Z;
+
+        // R. Byers. Solving the algebraic Riccati equation with the matrix sign
+        // function. Linear Algebra Appl., 85:267–279, 1987
+        // Added determinant scaling to improve convergence (converges in rough half
+        // the iterations with this)
+        const double ck = std::pow(std::abs(m_pimpl->Z.determinant()), -1.0 / p);
+        m_pimpl->Z *= ck;
+        m_pimpl->Z = m_pimpl->Z - 0.5 * (m_pimpl->Z - m_pimpl->Z.inverse());
```

Comment:
The same as above here. Almost surely, ``- m_pimpl->Z.inverse()`` generates a temporary object.

Reply:
It should be done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/157/commits/b3209960b6d9c6109d2ee92b2dad4638f74e5ae5